core/embed/rust/src/ui/model_mercury/component/button.rs

```diff
@@ -0,0 +1,440 @@
+                    State::Initial | State::Disabled => {
+                        // Do nothing.
+                    }
+                    State::Pressed if touch_area.contains(pos) => {
+                        // Touch finished in our area, we got clicked.
+                        self.set(ctx, State::Initial);
+                        return Some(ButtonMsg::Clicked);
+                    }
+                    _ => {
+                        // Touch finished outside our area.
+                        self.set(ctx, State::Initial);
+                        self.long_timer = None;
+                    }
+                }
+            }
+            Event::Timer(token) => {
+                if self.long_timer == Some(token) {
+                    self.long_timer = None;
+                    if matches!(self.state, State::Pressed) {
+                        #[cfg(feature = "haptic")]
+                        play(HapticEffect::ButtonPress);
+                        self.set(ctx, State::Initial);
+                        return Some(ButtonMsg::LongPressed);
+                    }
+                }
+            }
+            _ => {}
+        };
+        None
+    }
+
+    fn paint(&mut self) {
+        let style = self.style();
+        self.paint_background(style);
+        self.paint_content(style);
+    }
+
+    #[cfg(feature = "ui_bounds")]
+    fn bounds(&self, sink: &mut dyn FnMut(Rect)) {
+        sink(self.area);
+    }
+}
+
+#[cfg(feature = "ui_debug")]
+impl crate::trace::Trace for Button {
+    fn trace(&self, t: &mut dyn crate::trace::Tracer) {
+        t.component("Button");
+        match &self.content {
+            ButtonContent::Empty => {}
+            ButtonContent::Text(text) => t.string("text", *text),
+            ButtonContent::Icon(_) => t.bool("icon", true),
+            ButtonContent::IconAndText(content) => {
+                t.string("text", content.text.into());
+                t.bool("icon", true);
+            }
+            ButtonContent::IconBlend(_, _, _) => t.bool("icon", true),
+        }
+    }
+}
+
+#[derive(PartialEq, Eq)]
+enum State {
+    Initial,
+    Pressed,
+    Released,
+    Disabled,
+}
+
+#[derive(PartialEq, Eq)]
+pub enum ButtonContent {
+    Empty,
+    Text(TString<'static>),
+    Icon(Icon),
+    IconAndText(IconText),
+    IconBlend(Icon, Icon, Offset),
+}
+
+#[derive(PartialEq, Eq, Clone, Copy)]
+pub struct ButtonStyleSheet {
+    pub normal: &'static ButtonStyle,
+    pub active: &'static ButtonStyle,
+    pub disabled: &'static ButtonStyle,
+}
+
+#[derive(PartialEq, Eq)]
+pub struct ButtonStyle {
+    pub font: Font,
+    pub text_color: Color,
+    pub button_color: Color,
+    pub background_color: Color,
+    pub border_color: Color,
+    pub border_radius: u8,
+    pub border_width: i16,
+}
+
+#[derive(Clone, Copy)]
+pub enum CancelInfoConfirmMsg {
+    Cancelled,
+    Info,
+    Confirmed,
+}
+
+#[derive(PartialEq, Eq)]
+pub struct IconText {
+    text: &'static str,
```

Comment:
I would find it beneficial to have
```
pub struct IconText {
    text: TString<'static>,
    ...
```
because of menu items in `VerticalMenu`. 
We can do so in another PR though.

Reply:
bootloader currently doesn't use `VerticalMenu`, let's do it in #3585